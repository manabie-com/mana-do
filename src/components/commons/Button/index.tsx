import React, { forwardRef } from "react";
import cn from "classnames";
import "./style.css";

export enum ButtonSizes {
  lg = "lg",
  md = "md",
  xs = "xs",
}

export enum ButtonColors {
  primary = "primary",
  danger = "danger",
}

interface ButtonProps {
  disabled?: boolean;
  fullWidth?: boolean;
  type?: "submit" | "button" | "reset";
  onClick?: (event: React.SyntheticEvent) => void;
  color?: ButtonColors;
  className?: string;
  children?: React.ReactNode;
}

const defaultProps: ButtonProps = {
  disabled: false,
  fullWidth: true,
  type: "button",
  color: ButtonColors.primary,
};

const Button = forwardRef((props: ButtonProps, ref: any) => {
  const { color, className, children, disabled, fullWidth, type, ...rest } = {
    ...defaultProps,
    ...props,
  };

  const classOfComponent = cn(
    className,
    "btn-custom",
    `btn-custom-${color}`,
    disabled && "btn-custom-disabled"
  );

  return (
    <button
      className={classOfComponent}
      ref={ref}
      disabled={disabled}
      type={type}
      {...rest}
    >
      {children}
    </button>
  );
});

export default Button;
