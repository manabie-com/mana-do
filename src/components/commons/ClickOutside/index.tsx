import React, { useRef, useEffect } from "react";

/**
 * Hook that alerts clicks outside of the passed ref
 */
function useClickOutsideComponent(ref: any, onClickOutside: () => void) {
  useEffect(() => {
    /**
     * Alert if clicked on outside of element
     */
    function handleClickOutside(event: any) {
      if (ref.current && !ref.current.contains(event.target)) {
        onClickOutside();
      }
    }
    // Bind the event listener
    document.addEventListener("mousedown", handleClickOutside);
    return () => {
      // Unbind the event listener on clean up
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, [ref, onClickOutside]);
}

/**
 * Component that alerts if you click outside of it
 */
interface Props {
  onClickOutside?: () => void;
  children?: React.ReactNode;
}

function ClickOutsideComponent(props: Props) {
  const { onClickOutside = () => null, children } = props;

  const wrapperRef = useRef(null);
  useClickOutsideComponent(wrapperRef, onClickOutside);

  return <div ref={wrapperRef}>{children}</div>;
}

export default ClickOutsideComponent;
