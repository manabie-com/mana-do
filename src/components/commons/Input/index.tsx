import cn from "classnames";
import React, { forwardRef } from "react";
import "./style.css";

interface InputProps {
  className?: string;
  disabled?: boolean;
  readOnly?: boolean;
  type?: string;
  name?: string;
  id?: string;
  value?: string | number;
  onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
  onKeyDown?: (e: React.KeyboardEvent<HTMLInputElement>) => void;
  placeholder?: string;
  checked?: boolean;
}

const Input = forwardRef((props: InputProps, ref: any) => {
  const { value, className, disabled, ...rest } = props;
  const classOfRoot = cn("root-input", className);
  const classOfInputBase = cn(
    "input-custom",
    "input-size-md",
    disabled && "input-custom__disabled"
  );

  return (
    <div className={classOfRoot}>
      <input value={value} className={classOfInputBase} ref={ref} {...rest} />
    </div>
  );
});

export default Input;
