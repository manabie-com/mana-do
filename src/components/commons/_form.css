form label {
  width: 100%;
  padding-top: 10px;
  margin-top: 10px;
  display: flex;
  justify-content: center;
}
form label h2 {
  margin-right: 1em;
  display: flex;
  align-items: center;
  text-align: center;
}
form label input {
  min-width: 325px;
  height: 50px;
  border: solid 1px;
  border-radius: 25px;
  padding-left: 25px;
  font-size: 18px;
  font-weight: bold;
}
form label input:hover {
  border-color: yellow;
}


@media screen and (max-width: 480px) {
  form {
    font-size: 12px;
  }
  form label input {
    min-width: 200px;
    height: 30px;
    padding-left: 15px;
  }
  form .actions button {
    min-width: 150px;
    width: auto;
  }
}