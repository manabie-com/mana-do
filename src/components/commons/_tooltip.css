.use-tooltip {
  position: relative;
  display: inline;
  text-align: left;
  width: 100%;
}
  
.use-tooltip:hover .tooltip {
  display: block;
  z-index: 3;
  transform: translateX(-50%);
  display: -webkit-box;
  -webkit-line-clamp: 5;
  word-break: break-word;
  -webkit-box-orient: vertical;
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 0.5em;
  font-size: 16px;
  line-height: 24px;
  color: steelblue;
  font-weight: 500;
}

.use-tooltip:hover .tooltip.top {
  position: fixed;
  top: -20px;
  left: 50%;
  min-width: 80vw;
}

.use-tooltip:hover .tooltip.center {
  text-align: center;
}

.use-tooltip:hover .tooltip.left {
  text-align: left;
}

.use-tooltip:hover .tooltip.right {
  text-align: right;
}

.use-tooltip:hover .tooltip.current {
  
}
  
.tooltip {
  display: none;
  color: white;
  background-color: rgb(240, 247, 181, 0.7);
  margin: 2em;
  max-width: 80vw;
  border: 1px solid rgba(0, 0, 0, 0.4);
  border-radius: 5px;
  box-shadow: 4px 2px rgba(0,0,0, 0.4);
}

.typewriter {
}

.typewriter .tooltip-text {
  text-align: right;
  white-space: nowrap;

  /* animation properties */
  -moz-transform: translateX(-100%);
  -webkit-transform: translateX(-100%);
  transform: translateX(-100%);
  
  -moz-animation: running-text 20s linear infinite;
  -webkit-animation: running-text 20s linear infinite;
  animation: running-text 20s linear infinite;
}

/* for Firefox */
@-moz-keyframes running-text {
  from { -moz-transform: translateX(-100%); }
  to { -moz-transform: translateX(100%); }
}

/* for Chrome */
@-webkit-keyframes running-text {
  from { -webkit-transform: translateX(-100%); }
  to { -webkit-transform: translateX(100%); }
}

@keyframes running-text {
  from {
    -moz-transform: translateX(-100%);
    -webkit-transform: translateX(-100%);
    transform: translateX(-100%);
  }
  to {
    -moz-transform: translateX(100%);
    -webkit-transform: translateX(100%);
    transform: translateX(100%);
  }
}