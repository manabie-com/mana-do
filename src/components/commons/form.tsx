import React, { useCallback } from 'react'

export type FormProps = {
  model: any,
  formData: any,
  onSubmit: Function,
  onUpdateField: Function,
  submitText?: string
}

const Form = (props: FormProps) => {
  const { model, formData, onSubmit
    , onUpdateField, submitText
  } = props
  const fields = Object.keys(model).map((key) => {
    return model[key]
  })

  const handleSubmit = useCallback((e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault()
    onSubmit()
  }, [onSubmit])

  const handleSubmitClick = useCallback((e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault()
    onSubmit()
  }, [onSubmit])

  return (
    <form onSubmit={handleSubmit}>
      {fields.map((field) => {
        const { component, ...nestedProps } = field
        return component({ ...nestedProps, ...formData[field.name], onUpdateField })
      })}
      <div className='actions'>
        <button type='submit' onClick={handleSubmitClick}>
          {submitText}
        </button>
      </div>
    </form>
  )
}

export default Form