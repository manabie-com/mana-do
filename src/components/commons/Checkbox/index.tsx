import React, { forwardRef, useRef } from "react";
import cn from "classnames";
import CheckboxChecked from "./icon/CheckboxChecked";
import CheckboxUnChecked from "./icon/CheckboxUnChecked";
import Input from "../Input";
import "./style.css";

interface Props {
  className?: string;
  iconRef?: React.Ref<unknown>;
  disabled?: boolean;
  children?: React.ReactNode;
  checked?: boolean;
  value?: string;
  name?: string;
  onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

const defaultProps: Props = {
  disabled: false,
  checked: false,
};

const Checkbox = forwardRef((props: Props, ref: any) => {
  const {
    className,
    iconRef,
    name,
    disabled,
    checked,
    value,
    children,
    onChange,
    ...rest
  } = {
    ...defaultProps,
    ...props,
  };

  const classOfRoot = cn(
    "checkbox-custom",
    className,
    disabled && "checkbox-custom-disabled",
    checked && "checkbox-custom-checked"
  );

  const refOfInput = useRef();

  return (
    <label
      data-testid="checkbox-custom"
      className={classOfRoot}
      ref={ref}
      {...rest}
    >
      {checked ? (
        <CheckboxChecked ref={iconRef} />
      ) : (
        <CheckboxUnChecked ref={iconRef} />
      )}

      {children}
      <Input
        readOnly
        type="checkbox"
        ref={refOfInput}
        name={name}
        value={value}
        className="checkbox-custom-input"
        checked={checked}
        onChange={onChange}
      />
    </label>
  );
});

export default Checkbox;
