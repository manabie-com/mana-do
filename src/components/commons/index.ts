import Input from "./Input";
import Button, { ButtonColors } from "./Button";
import PrivateRoute from "./PrivateRoute";
import Paper from "./Paper";
import Checkbox from "./Checkbox";
import Modal from "./Modal";
import ModalConfirm from "./ModalConfirm";
import ClickOutsideComponent from "./ClickOutside";
export {
  PrivateRoute,
  Input,
  Button,
  ButtonColors,
  Paper,
  Checkbox,
  Modal,
  ModalConfirm,
  ClickOutsideComponent,
};
