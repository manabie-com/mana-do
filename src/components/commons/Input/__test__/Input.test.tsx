import React from "react";
import ReactDOM from "react-dom";
import { render, cleanup, fireEvent } from "@testing-library/react";

import { Input } from "components/commons";
import "@testing-library/jest-dom";
import renderer from "react-test-renderer";

const propsTest = {
  onChange: jest.fn(),
  value: "Value test",
  className: "Input-Custom",
};

let container: HTMLElement;

beforeEach(() => {
  container = render(<Input {...propsTest} />).container;
});

afterEach(cleanup);

test("Input: render without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<Input {...propsTest} />, div);
  ReactDOM.unmountComponentAtNode(div);
});
test("Input: props of input", () => {
  const rootEl = container.querySelector(".root-input") as HTMLDivElement;
  const inputEl = container.querySelector(".input-custom") as HTMLInputElement;
  expect(rootEl?.className).toContain(propsTest.className);
  expect(inputEl?.value).toEqual(propsTest.value);
});
test("Input: onChange call when input value", () => {
  const inputEl = container.querySelector(".input-custom") as HTMLInputElement;
  fireEvent.change(inputEl, { target: { value: "test" } });
  expect(propsTest.onChange).toBeCalled();
});
test("Input: match snapshot", () => {
  const tree = renderer.create(<Input {...propsTest} />).toJSON();
  expect(tree).toMatchSnapshot();
});
