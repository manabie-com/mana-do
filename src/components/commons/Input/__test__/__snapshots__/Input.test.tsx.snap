// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input: match snapshot 1`] = `
<div
  className="root-input Input-Custom"
>
  <input
    className="input-custom input-size-md"
    onChange={
      [MockFunction] {
        "calls": Array [
          Array [
            SyntheticEvent {
              "_dispatchInstances": null,
              "_dispatchListeners": null,
              "_targetInst": null,
              "bubbles": null,
              "cancelable": null,
              "currentTarget": [Function],
              "defaultPrevented": null,
              "dispatchConfig": null,
              "eventPhase": null,
              "isDefaultPrevented": [Function],
              "isPropagationStopped": [Function],
              "isTrusted": null,
              "nativeEvent": null,
              "target": null,
              "timeStamp": [Function],
              "type": null,
            },
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    value="Value test"
  />
</div>
`;
